docs(sampling): rewrite sampling docs for practical usage (#515)
modelcontextprotocol/modelcontextprotocol: docs/docs/concepts/sampling.mdx

```typescript
{
  capabilities: {
    sampling: {}
  }
}
```

## Implementing sampling

Sampling can be a standalone server-initiated interaction, but works best when used within other MCP features. The following example demonstrates how sampling can be used within tool interactions:

### Server

```typescript
import assert from "node:assert";
import { z } from "zod";

const server = new McpServer({
  name: "example-server",
  version: "1.0.0",
});

// Add an addition tool
server.tool(
  "add",
  { a: z.number(), b: z.number() },
  async ({ a, b }, extra) => {
    // Send the client a sampling request and wait for a response
    const result = await extra.sendRequest(
      {
        method: "sampling/createMessage",
        params: {
          messages: [
            {
              role: "user",
              content: {
                type: "text",
                text: `Add ${a} and ${b}. Respond with only the sum of the two numbers.`,
              },
            },
          ],
          maxTokens: 1000,
          systemPrompt: "You are very good at math.",
          includeContext: "thisServer",
        },
      },
      CreateMessageResultSchema

>     );

Reviewer: Out of scope for this PR, but we really need a friendlier API. :laughing:

Author: Yeah, honestly even the name of the feature ("sampling") doesn't really convey what this is used for, let alone the API itself 😔

Reviewer: Agreed!  If we could unify sampling and elicitation, I would propose the term "solicitation".

Quoth Gemini 2.5 Pro on "elicit" vs "solicit":

> * To **solicit** a response is to _request_ it.
> * To **elicit** a response is to _get_ it (often without a direct request for that specific response).
>
> In essence, "solicit" is about the asking, while "elicit" is about the succeeding in drawing something out.

Reviewer 2: I have to disagree that using two very similar sounding words instead will clarify things. :)

Reviewer 2: @jonathanhefner thanks, I totally misread the comment.

    },
  }
);

async function sendPrompt(_prompt: string): Promise<string> {
  // In a real application, you would write code to make a request to an LLM
  return "3";
}

// Define a request handler that determines how the client should respond to sampling requests
client.setRequestHandler(CreateMessageRequestSchema, async (request) => {
  // In a real application, you would raise a prompt to a user before handling the request
  // ...

  // Send the prompt to a model
  const prompt = `${
    request.params.systemPrompt
  }\n\n${request.params.messages.join("\n\n")}`;
  const modelResult = await sendPrompt(prompt);

  // Return the response to the server
  return {
    model: "my-model",
    role: "assistant",
    content: {
      type: "text",
      text: modelResult,
    },
  };
});

await client.connect(transport);

const toolResult = await client.callTool({
  name: "add",
  arguments: {
    a: 1,
    b: 2,
  },
});
```

**Result:**

```typescript
{ content: [ { type: 'text', text: 'The sum of 1 and 2 is 3.' } ] }
```

## Example sampling patterns

Here are some examples of how servers can use sampling to enhance interactions:

### Natural-language tool responses

> The implementer of a server without direct LLM access may want a tool to return the result of an operation in natural language to aid clients:

Reviewer: Maybe rephrase for clarity?

```suggestion
A tool without direct LLM access may want to perform an operation using natural language:
```

Author: Good suggestion, committed
